src/devices/machine/swtpc8212.cpp

```diff
@@ -0,0 +1,475 @@
+	: swtpc8212_device(mconfig, SWTPC8212, tag, owner, clock)
+{
+}
+
+WRITE8_MEMBER(swtpc8212_device::latch_w)
+{
+	// Bits 0 to 3 control outputs that are intended to control and tape
+	// 'read' and 'punch' operations. These are strobes, about 10usec, and
+	// intended to trigger on the falling edge.
+	//
+	// Bit 0 - Read on output
+	// Bit 1 - Read off output
+	// Bit 2 - Punch on output
+	// Bit 3 - Punch off output
+
+	// Bit 4 - ?? usually high
+	// Bit 5 - ?? usually low
+
+	// Bits 6 and 7 change with the screen format.
+	//
+	// Bit 6 is zero for formats I and II, and one for formats III and IV
+	// and for the graphics format. Assume this selects between the
+	// characters sets, so formats III and IV might use an alternative
+	// custom text character set.
+	//
+	// Bit 6 - character set: 0 - standard; 1 - alternate/graphics.
+	//
+	// Bit 7 is zero in formats I, II, III, and IV, and one for the
+	// graphics format. Assume this controls the horizontal
+	// inter-character gap, eliminating the gap for the graphics format.
+	//
+	// Bit 7 - character width: 0 - 9 dots; 1 - 8 dots.
+	//
+	if (BIT(data, 7) == 0)
+		m_crtc->set_char_width(9);
+	else
+		m_crtc->set_char_width(8);
+
+	m_latch_data = data;
+}
+
+
+READ8_MEMBER(swtpc8212_device::pia0_pa_r)
+{
+	// PA0 controls the 'duplex' mode, the echoing back of characters, and
+	// appears to connect to a switch on the outer casing.
+	//
+	// PA1 On the CT-82 this enabled or disabled use of an optional ROM,
+	// but that function is disabled in the 8212, and this is probably
+	// unused.
+	//
+	// PA2 is Jumper B, and PA3 is Jumper A.
+	uint8_t config = m_config->read();
+
+	// TODO:
+	// PA7 - cursor bit 0
+	// PA6 - cursor bit 1
+	// PA5 - cursor bit 2
+	// PA4 - cursor bit 3
+
+	return config;
+}
+
+READ8_MEMBER(swtpc8212_device::pia0_pb_r)
+{
+	return m_kbd_data;
+}
+
+void swtpc8212_device::kbd_put(uint8_t data)
+{
+	m_kbd_data = data;
+	// Triggers on the falling edge.
+	m_pia0->cb1_w(ASSERT_LINE);
+	m_pia0->cb1_w(CLEAR_LINE);
+	m_pia0->cb1_w(ASSERT_LINE);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::pia0_ca2_w)
+{
+	if (state == 0)
+	{
+		m_beeper->set_state(1);
+		m_bell_timer->reset(attotime::from_msec(250));
+	}
+}
+
+WRITE8_MEMBER(swtpc8212_device::pia1_pa_w)
+{
+	// External parallel printer data output.
+	m_printer_data = data;
+}
+
+READ_LINE_MEMBER(swtpc8212_device::pia1_ca1_r)
+{
+	// External parallel printer busy input.
+	return 0;
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::pia1_ca2_w)
+{
+	// External parallel printer data ready.
+
+	// Trigger on the falling edge.
+	if (m_printer_data_ready == 1 && state == 0)
+	{
+		m_printer->output(m_printer_data);
+		// Toggle the printer busy line as the software waits for a
+		// falling edge.
+		m_pia1->ca1_w(CLEAR_LINE);
+		m_pia1->ca1_w(ASSERT_LINE);
+		m_pia1->ca1_w(CLEAR_LINE);
+	}
+	m_printer_data_ready = state;
+}
+
+READ8_MEMBER(swtpc8212_device::pia1_pb_r)
+{
+	return m_dip_switches->read();
+}
```

Comment:
Why do you have this member here at all?  You can use `.set_ioport(...)` to connect a devcb directly to an I/O port, and avoid the trampoline.